### ISSUE TITLE
Subscriptions - Save file?

### ISSUE THREAD

**Issue by nojohnso** [FreeTubeApp/FreeTube#370]
If this doesn't exist already, I'd love to see a subscription file that was always updated with any changes and the ability to save to a chosen location. It would make for an easy way to keep subs synced across devices using something like ownCloud.

**Comment by PrestonN** [comment]
The plan is to eventually support the WebDAV protocol which will allow syncing with various cloud services including NextCloud/ownCloud and Dropbox.  There isn't any ETA on it but it will likely be the first thing I work on after a stable release (maybe before, not sure yet).

For now, I've heard of a few cases where users have had success using Syncthing between different systems.  You can sync the folder at #151 for your respective systems and should see it update across devices.

**Comment by Rik44444** [comment]
just to say a webdav sync function would be awesome!

**Comment by microraptor** [comment]
WebDAV seems like the most elegant solution for this. Then perhaps an integration for NewPipe could also be made as discussed in #1026. I just want to point out that the user @T-vK already started developing this a while back, but doesn't seem to have opened a PR for it:
https://github.com/TeamNewPipe/NewPipe/issues/1163#issuecomment-713482853
The code he wrote:
https://github.com/T-vK/FreeTube/commit/1df250538cda1163bcdd0cc6f2d8731b869ff483

**Comment by hilburger** [comment]
Oh please, I'd love to see this working, too!  :-D 
This is one one the last parts for me to get rid off Google Services on phones and computers at home <3

**Comment by therealrobster** [comment]
Just chiming in, sync would be amazing.  NewPipe to FT and FT to FT.

**Comment by ChunkyProgrammer** [comment]
https://www.npmjs.com/package/webdav seems like a good option. Would need to find a way to keep passwords secure and not stored as plaintext on the user's system or only use oAuth

**Comment by maplepy** [comment]
can't wait for this!

**Comment by BBArikL** [comment]
WebDAV sync would be great! :+1:

**Comment by absidue** [comment]
@simmstein I would highly recommend removing the watcher while FreeTube is running unless you want your files to get corrupted. While FreeTube is running it is meant to be the only program accessing the files and while it is running the files should be considered incomplete. While FreeTube is running it uses an append-only format to avoid having to waste disk writes rewriting the whole file every time, only when all FreeTube windows have been closed and it has finished cleaning up the databases should they be considered clean and syncable. With your sync while running approach, if you were to have FreeTube open on multiple computers at the same time, you would be very likely to encounter dataloss or odd behaviour from FreeTube (if the databases are too corrupted the app intentionally refuses to start).

We really need to find a way to lock the files like other programs for, so people stop messing around with the files while FreeTube is running.

**Comment by simmstein** [comment]
The sync is clever than just dectect an update and push to the server. And databases are never rewrite when freetube is running because data is Pull is processed before freetube start.

**Comment by absidue** [comment]
I can stop you from doing what you want to do, but please make sure it is clear to your users that if **anything** happens to their databases while using your tool they should go to you for support, not us (FreeTube team).

Example of a case that could cause problems:
1. Start FreeTube on computer 1
2. Make some changes
3. Watcher pushes changes to the server while FreeTube is running which means it's incomplete
4. Start FreeTube on computer 2
5. Computer 2 pull incomplete file from server

It would be better if the sync tool worked like this:
1. Pull data from server
2. Start FreeTube
3. Sync tool watches the FreeTube processes
4. When there are no more FreeTube processes running it sends the files to the server

**Comment by simmstein** [comment]
My tool does not save plain file, it's a sqlite database. Pushed datas are validated in client and server side. I also analyse how freetube files are updated while freetube is running or stoppe. It does not mean that it's perfect but corrupted data seems quite impossible.

**Comment by absidue** [comment]
As mentioned above, if any user have any issues using their tool, make sure they know they need to come to you. Add a disclaimer to your README or something like that please.

The only officially supported way to sync your data between FreeTube installs is exporting and importing in the Data Settings in FreeTube.

**Comment by bingoxo** [comment]
@simmstein can you please make an android fork for your app to work with newpipe and/or FreeTubeAndroid

**Comment by Skyedra** [comment]
Would love to be able to do cross-device sync, particularly via SyncThing.

I can see webdav support being useful, but I'd like to suggest a JSON / plain text file synchronization target as well so it can work with non-cloud sync options (such as SyncThing).

I do see there is a workaround referenced above (#151), but it seems like syncing the database files would have the same concurrency risks as simmstein's method...?  Additionally, resolving sync conflicts with binary files tends to be quite painful.

**Comment by flupkede** [comment]
I personally don't see the need for webdav, it's not the greatest protocol around. I use Resilio sync and if one would use dropbox or syncthing it could all work, if there would be a way that freetube could handle the overwrite of json files (an update coming from a different source) whilst running.

**Comment by Skyedra** [comment]
Also, it might be best for each freetube instance to write to its own json inside a particular folder, and the active freetube instance reads/watches all json files in the target folder to look for changes.  This way there would not be sync conflicts if two freetube instances are watching things at the same time (ie: music playing in living room while watching content in office, etc), or devices get disconnected from network temporarily, etc.

**Comment by guillermodotn** [comment]
# Possible implementation

Moved here from issue #6788, as it was closed due to issue duplication.

### Problem Description

To enhance user experience for those utilizing FreeTube on multiple devices, I propose implementing a feature that synchronizes watch history and progress across different FreeTube instances.

### Proposed Solution

This can be achieved by leveraging Git, a distributed version control system, to manage and synchronize the relevant data.

## Implementation Steps:

1. **Private Repository Setup:**
  Users create a private repository on a Git hosting service of their choice, such as:
    - GitHub
    - GitLab
    - Gitea
    - ...

2. **Data Pull on Startup:**
Upon launching FreeTube, the application pulls the latest data from the user's designated repository to update watch history and progress.

3. **Data Push on Exit or Triggered Action:**
When closing FreeTube or upon a user-initiated action, the application commits any changes to the watch history or progress and pushes these updates to the repository.

## Advantages:

* **No Need for Additional Infrastructure:**
This approach eliminates the requirement for new authentication systems or user management infrastructure, as it utilizes existing Git platforms.

* **Alignment with FreeTube's Principles:**
By adopting this method, FreeTube continues to uphold its commitment to customization, privacy, and user freedom.

### Alternatives Considered

Currently, users can manually export their watch history from one FreeTube instance and import it into another. This process involves exporting the history file from the source device and importing it on the target device via the application's settings. 

### Additional Information

This proposal is inspired by practices used to synchronize configuration files (dot-files) across multiple systems, where Git repositories are employed to maintain consistency and version control.

**Comment by Skyedra** [comment]
I don't think most users would want to have to manually sort out git merge conflicts...  They will want it to "just work"

It is better to have each device write to its own file as I described above, because:

1) No merge conflicts
2) The sync system can be independent to the storage (syncthing, NFS, dropbox, etc)
3) In the case of two devices watching a video and needing to decide which data to use, Freetube can resolve this easily by checking a "last modified" datetime field for each video.  This will reveal which data was more recent without the user having to manually merge files.

**Comment by guillermodotn** [comment]
I see your point, and yes, I would prefer something more seamless. The issue I see with synchronization tools like Syncthing or NFS is that they require the devices to be on at the same time, and most likely, they need to be on the same network, depending on your setup.

Regarding merge conflicts, I don’t see it as an issue since the repo will be pulled at the start of the application to always maintain a linear history. In the event of a conflict (e.g., two devices syncing at the same time), it can be force-pulled based on the last modification date.

As for third-party cloud storage solutions like Dropbox or Box, I’m more inclined toward those, as more people are familiar with how to use these services.

**Comment by absidue** [comment]
If you want to use git, you can already do that yourself without requiring any changes in FreeTube. The data location and the relevant files are documented here https://docs.freetubeapp.io/usage/data-location/. So doing it manually you would just have to git pull before launching FreeTube, then after you have closed the last window and waited for the FreeTube processes to vanish from the task manager (it does some cleanup on the caches and databases after the last window has closed), you would commit and push. I would highly discourage trying to sync the files while FreeTube is running, as it assumes it has exclusive access and control over the files while it is running.

**Comment by absidue** [comment]
As for 3rd party storage providers if any of them were to ever get integrated into FreeTube they would have to be fully open source and self-hostable. Those are the same criteria we have for all other integrations which is why we have SponsorBlock and DeArrow integrations but not ReturnYouTubeDislikes as they don't provide database dumps so it's not realistically self-hostable.

**Comment by tkreuzer** [comment]
The most simple way is this:
- Add a setting to allow specifying the subscription/playlist storage on network share / cloud synced folder
- Load the file on startup and update on every modification
- When updating, first reload the file, add the update, save the file
- To prevent synchronization issues, each access operation is "locked" by first creating a "settingfile.lock" file. If the lock file is successfully created, the actual file can be accessed, otherwise wait a bit and retry a few times.

**Comment by eumario** [comment]
Looking at this, as something I want for all the various YouTube Alternative Front Ends.  I'm wondering, if not just creating a simple REST api would work best for this, to store/fetch watch history, and time codes for videos, through a Simple self-hosting server that stores said data into a SQLite database, wouldn't make things 100 times easier.  It can even be done up as a simple Docker image, for easy spin up of this kind of information.  At most, it would just need the Video ID, a bool to determine if the video is fully watched / marked as watched, and then a int for the timestamp of watched to.  No need to over-complicate things with syncs, and such.

**Comment by eumario** [comment]
After doing some quick mock ups, and reviewing the *.db files in question, I created a simple C# project that acts as a Single User Sync storage to store everything from History, Settings, Playlists, Profiles, Search History.  But due to how late it is, I have only tested and verified History and Settings is working as intended.  I still have to test Playlists, Profiles and Search Histories.  You can check out the code here: https://forge.eumario.dev/eumario/freetubesync/

**Comment by eumario** [comment]
I have fully tested the API, with a couple of tweaks (due to mistakes from being exhausted), the Sync Storage API has been finished.  I'm currently working on a Tray App client to sync the database files, with the API.   Doing this, will not be optimal, but will be proof of concept, till the API can be implemented in the project itself.  It will monitor the files for changes, and sync them with the storage database.  The down side, is that it can't update the database files with the App open.  So the app needs to be closed, in order for the sync client to update from the Storage API.  Once updated, the App will have everything synced between multiple instances.  (Everything except the Subscription Cache database).

**Comment by therealrobster** [comment]
This looks really interesting I will definitely take a look.  Thank you so much for contributing, hopefully it can work for what people need! 

Update, I try to follow the link but I get a `Gateway Timeout` error.

**Comment by eumario** [comment]
> This looks really interesting I will definitely take a look. Thank you so much for contributing, hopefully it can work for what people need!
> 
> Update, I try to follow the link but I get a `Gateway Timeout` error.

Sorry, the ISP where I'm hosting it, was doing some updates.  It's up right now, so you should be able to get to it.

**Comment by eumario** [comment]
I've done some more updates to the rest API, and I have created a Cross-Platform System Tray App, to synchronize things with the Rest API and the local install.  The Rest API has a Docker Release, where you can install it on any Docker compatible setup for holding the sync data, and the Client can be found here:  https://forge.eumario.dev/eumario/FreeTubeSyncer It's a Dotnet 9 project, and I will see about packing it for all three platforms, but you can clone it, and build it with Dotnet.

The syncer cannot sync from the REST API to the FreeTube database files while FreeTube is open, as the changes won't be reflected in the App, so it monitors for when the app is closed, then synchronizes the stuff fetched from the REST Api to the client.  It will detect changes that are made to the FreeTube database files automatically (Even while FreeTube is running), and send them off to the server.

It's not an Optimal solution, but it's a proof of concept, and shows how the REST API server can be directly integrated with FreeTube, and has a client setup to allow the synchronization until such time as it's implemented, or not.

**Comment by yosefario-dev** [comment]
Is the feature still in demand?
I think on trying to integrate the sync as a native feature into the FreeTube app.
I was thinking first of P2P with encrypted blobs, and maybe a serverto enable async sync, but I see that people want cloud.
Never worked with cloud storage APIs, but I can try later.
P2P is 100% possible and will be not super hard.

**Comment by Nama** [comment]
I'd prefer ssh to an existing server, but I also understand that not everyone has a server at hand.

**Comment by yosefario-dev** [comment]
> I'd prefer ssh to an existing server, but I also understand that not everyone has a server at hand.

Thats exactly what I tried accounting for, but after 3 hours of trial and error and research, I see that server-less P2P is very unstable and not always works.
So I will instead make server-based syncing, but in a way that even public/community instances can be trusted for those who cant self-host.
Cause if the solution works only for server owners who trust their server because they own it, it will not be exactly a feature worth integrating to everybody.
If its integrated into the app, then it should be usable for most users no matter if they own a server or not.
But in the same time it should not compromise on privacy.
I am currently working on it. Planning stage.

Edit (10.08.25 19:34 UTC+0): Planning stage is finished. I will get onto making the code. Hopefully the feature will be ready soon.

Edit (10.08.25 21:15 UTC+0): Done with implementing the UI within the native settings menu. Keeping same style. Ongoing onto functionality development stage.

Preview:
<img width="1485" height="845" alt="Image" src="https://github.com/user-attachments/assets/83184c68-b90a-4b00-81bb-d474bd818b88" />


Edit (11.08.25 23:10 UTC+0): Revised the UI. Added functionality to saving the config. Saved and persistent across reloads and restarts: Passphrase, Data Sync preferences, Sync timing preferences, Frequency of auto syncs and server URL.

Preview:
Default Settings View:
<img width="1639" height="938" alt="Image" src="https://github.com/user-attachments/assets/a7f79f1a-b82f-42ae-9879-059a7c6b0f2e" />

Advanced Settings View:
<img width="1639" height="938" alt="Image" src="https://github.com/user-attachments/assets/b3c6ec52-11a2-40d0-b780-8f1a20b9e465" />

**Comment by yosefario-dev** [comment]
I need to work now on extraction of the data: Profiles, Subscriptions, Watch History, Playlists and Application settings.
Then, I will setup the encapsulation and encryption per standards.
But that's the task for tomorrow. I sure do hope feature comes in handy to people.

**Comment by yosefario-dev** [comment]
I will use Web Crypto API (as it is built-in and will work without extra dependencies.) with AES-256-GCM encryption to encrypt the data at transfer to the server.
I dont want to give an ETA, but I think feature can be pushed in the next couple days after I finish the work and polish it.
Add proper translation keys, and etc to comply with the contribution guidelines.
Per my check, work it about 40% done.

**Comment by yosefario-dev** [comment]
Test of imports and exports of data were successful.
Encryption is successful.
I will develop the server and possibly the feature will be finished tomorrow.
I will link the repo with the server code.
Main Idea is to have 2 modes: Public and Private.
This way I can enable more features on private servers in cost of leaving more metadata.
"Public mode" is designed to leave minimum metadata, so it cant be abused by malicious server hosts.

I will host the first instance myself, in the public mode, at: https://sync.techvoid.co which will be set as default in settings.
Later on I think there should be a community-maintained list of public instances that can be fetched automaticly.
After server is done, I will finish the API side of the feature and then after some polish make a PR.

Current UI:
<img width="1632" height="985" alt="Image" src="https://github.com/user-attachments/assets/6cd690d7-4316-46cc-a3f3-5bfa1ed65789" />

<img width="1632" height="985" alt="Image" src="https://github.com/user-attachments/assets/fba6ee83-5d8d-401f-b393-e0c966e81d3f" />

<img width="1632" height="985" alt="Image" src="https://github.com/user-attachments/assets/7f03f6f9-06b6-445f-bb76-283b2fa0fed8" />

<img width="1632" height="985" alt="Image" src="https://github.com/user-attachments/assets/6081d9d2-ce2c-40e2-9a02-60ac8019f342" />

<img width="1632" height="1029" alt="Image" src="https://github.com/user-attachments/assets/ad9ebb45-f518-4cbb-b73b-3e339e8339bd" />

<img width="1632" height="1029" alt="Image" src="https://github.com/user-attachments/assets/5bf2d0ba-d6de-4849-95bd-894b4021c5ab" />


My current solution/protocol is made specifically to use with public instances.
Later on I suppose the feature will be expanded onto more protocols, as the community desires.
For example. the mentioned cloud services and etc.
But right now, I will stick with the paranoid protocol.

**Comment by absidue** [comment]
> I will host the first instance myself, in the public mode, at: https://sync.techvoid.co which will be set as default in settings.

If that is your plan for it, I will definitely veto the PR. Having it default to uploading user data to a server run by an unknown random on the internet that isn't even affiliated with the FreeTube team, so there is no trust or way to verify that you are not doing it for malicous purposes is a big no no. It would have to be entirely self-hosted.

**Comment by absidue** [comment]
Just a reminder that even submitting SponsorBlock segments from inside FreeTube is considered too privacy invasive to implement and SponsorBlock is run by someone that is actually known, so uploading people's entire user data to some random's server is obviously completely out of the question.

**Comment by yosefario-dev** [comment]
> If that is your plan for it, I will definitely veto the PR. Having it default to uploading user data to a server run by an unknown random on the internet that isn't even affiliated with the FreeTube team, so there is no trust or way to verify that you are not doing it for malicous purposes is a big no no. It would have to be entirely self-hosted.

Noted. I will remove the default instance.
Any other things I should keep in mind?

**Comment by yosefario-dev** [comment]
@absidue 
I think there should be at least one public instance, which will be set as default.
Without a public instance, casual users who have no exp. in self-hosting won't be able to use the feature.
It doesn't matter who runs it, so if it makes things better, any of the official/trusted members of FreeTube team can host the community server for everybody to use, and add it into the app as default.

> Having it default to uploading user data to a server run by an unknown random on the internet that isn't even affiliated with the FreeTube team, so there is no trust or way to verify that you are not doing it for malicous purposes is a big no no.

Server doesn't host any actual data. It is all E2EE. With minimal metadata. Even Filenames/FileIDs are made to be random to anybody who doesn't know the secret. The metadata is very limited and is unusable by server maintainer even if they had malicious intent.
There is no need to trust anyone with your data, and that is the best part.

**Comment by absidue** [comment]
You are welcome to maintain a custom build of FreeTube for yourself with that feature if you don't see it worth adding without a public instance.

**Comment by yosefario-dev** [comment]
> You are welcome to maintain a custom build of FreeTube for yourself with that feature if you don't see it worth adding without a public instance.

Thats not what I meant lmao.
But okay.
Just a question, why are you being so critical of the idea?
I dont see any reason to not have at least one community server for everybody to use.

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
We are being critical of everything, especially when it comes to our users privacy

**Comment by yosefario-dev** [comment]
Got it. Thanks.

**Comment by absidue** [comment]
I am not critical of the idea of having some kind of syncing functionality (ideally directly between two devices on the same local network), I am critical of the fact that you think the only viable way of doing it is having lots of users upload their data to some random server that will be set as the default, so most people won't change it. It goes the against the privacy principles that we've been trying to stick to with FreeTube (connect to as few 3rd party services as possible and when we do, upload as little data to them as possible). One of the major points of FreeTube is that your user data is stored locally.

In addition to the proposed functionality not fitting with FreeTube goals, the UI in the screenshots doesn't match the rest of Freetube so that would have to be redesigned (the screenshots look like you've copy pasted part of another app into FreeTube).

**Comment by yosefario-dev** [comment]
>I am not critical of the idea of having some kind of syncing functionality (ideally directly between two devices on the same local network), I am critical of the fact that you think the only viable way of doing it is having lots of users upload their data to some random server that will be set as the default, so most people won't change it. It goes the against the privacy principles that we've been trying to stick to with FreeTube (connect to as few 3rd party services as possible and when we do, upload as little data to them as possible). One of the major points of FreeTube is that your user data is stored locally.

Thanks for the answer.
I came to this solution and found it as most reliable in cases when devices are not on the same network.
If both devices are on a local network, no server is needed to sync between them. The P2P issues I described were only in cross-network scenarios.
Same-network sync via P2P is easy to implement. Thanks for reminding me to include that too.

>In addition to the proposed functionality not fitting with FreeTube goals, the UI in the screenshots doesn't match the rest of Freetube so that would have to be redesigned (the screenshots look like you've copy pasted part of another app into FreeTube).

About functionality: I understand your concern.
As for the UI, I just wanted it to look a bit modern. I will redesign it asap based on the native CSS.

To summarize, so I know what I should do, I should:
- Remove default server and prompt users for custom URL, always.
- Redesign the UI to fit the styling of the app.
- Add local network P2P  sync.

Can you please correct me if I am wrong, or should I go ahead with this?

I just try to contribute in what I can.

**Comment by absidue** [comment]
Ideally the syncing would happen locally, either directly between devices on a local network or with a server thing that is designed for use on a local network, not over the internet. The idea that syncing data between two devices on the same local network requires sending the data out over the internet just seems wrong, even more so in a privacy friendly app, especially when one of the advertised features of FreeTube is that your data is local on your device and not somewhere on the internet.

**Comment by yosefario-dev** [comment]
> Ideally the syncing would happen locally, either directly between devices on a local network or with a server thing that is designed for use on a local network, not over the internet. The idea that syncing data between two devices on the same local network requires sending the data out over the internet just seems wrong, even more so in a privacy friendly app, especially when one of the advertised features of FreeTube is that your data is local on your device and not somewhere on the internet.

Then I understood you right.
Because syncing across a local network can be done without a server.
It is not used at all if both devices are on the same network.
My proposed solution only requires a server accessible over internet only for syncing of devices which are not in same network.

From your reply, it seems to me as if you thought that for syncing across local network with my solution you would be required to have a server, which is not the case.

**Comment by absidue** [comment]
> From your reply, it seems to me as if you thought that for syncing across local network with my solution you would be required to have a server, which is not the case.

No, I am saying either send it directly between the devices on the local network for syncing or design some local network only server that can sync between the devices on the local network, which could act as a backup solution but also allowing you to sync asynchronously e.g. you use device 1, it sends stuff to the local server, you stop using it and a few hours later start using device 2, that fetches from the local sever and sends its own changes back.

Syncing over the internet should not be part of FreeTube unless we all decide that we want to give up on one of the core ideas of FreeTube that your data is only stored locally.

**Comment by absidue** [comment]
Additionally any syncing solution you come up with has to work for both the Electron and Android versions of FreeTube.

**Comment by yosefario-dev** [comment]
Oh. Okay.
Thanks. I will focus on a local network solution then, with optional server.
As for Android compatibility, I am using only vanilla JS and native API that exist in browsers (Electron is Chromium), so I suppose there shouldn't  be any problems.
Is there any special things to consider to ensure Android compatibility?
(Until this moment I didn't know that there is Android version of FreeTube, as I have seen that you didn't plan to create it.)

**Comment by Skyedra** [comment]
yosefario, I'd point you to my comments above for a solution that requires no server at all (local or cloud) to do syncing

**Comment by BeatLink** [comment]
Ok, so I think a quick and easy way we could get this implemented is by changing the backend database to a replicating database like pouchDB. Freetube already uses a NoSQL database format, same as pouchDB. PouchDB is also designed to run in-browser with a small footprint, which would make it great for FreeTube. 


All we would need to do is recode the backend, add some UI settings to configure sync with another server and implement a conflict management system (maybe by showing the user two conflicting actions and asking them which one is the winning one). I think that's the easiest and fastest way to implement freetube syncing. Bonus point is its private and secure. Devices can sync directly with each other without an intermediary server. And if persons wanted to host a central server, all they would need to do is spin up a CouchDB instance.

**Comment by squinonescolon** [comment]
I would just like automatic export/backup of subs and lists. My computer crashed and I lost a lot of subscriptions and lists and I would love to back up frequently without having to remember.

**Comment by aldoushuxxx** [comment]
> I would just like automatic export/backup of subs and lists. My computer crashed and I lost a lot of subscriptions and lists and I would love to back up frequently without having to remember.

YES. I am not even looking for a full synch, but really just a way to export the current sub list on one device so I can copy it to my other device. CSVs would work.

**Comment by VewDev** [comment]
@eumario made what looks like a working REST API that could be used to sync user data. It does require a server, but the "user data being sent somewhere on the internet" issue can be fixed just by not providing a default instance. If a user wants to sync their data to a cloud, they would either have to self host it or find a trustable instance.

From their comment:

> Looking at this, as something I want for all the various YouTube Alternative Front Ends. I'm wondering, if not just creating a simple REST api would work best for this, to store/fetch watch history, and time codes for videos, through a Simple self-hosting server that stores said data into a SQLite database, wouldn't make things 100 times easier. It can even be done up as a simple Docker image, for easy spin up of this kind of information. At most, it would just need the Video ID, a bool to determine if the video is fully watched / marked as watched, and then a int for the timestamp of watched to. No need to over-complicate things with syncs, and such.

I totally agree a REST API would be the easiest and more future-proof way of doing this. It would also very easily allow syncing the data across different client applications (like NewPipe) just by integrating the API into the application.

Regarding restricting the feature to local network, this would not fit my (and I assume many others) use case where I would want to use FreeTube in mobile and will not be connected to the same networks at all times.

Right now eumario's API requires a separate client that uploads data to the server when FreeTube is closed, but I believe this could be integrated into FreeTube to upload modifications "real time" to the server (aka notifying the server when a new subscription is created, when a new video is watched...). This would completely remove the problem of different client instances syncing data at the same time, since the authority (the server) could just take the latest request in case of conflict (like watching the same video on 2 devices at once and stopping at different timestamps).

Read their comment thread for more info on the matter: https://github.com/FreeTubeApp/FreeTube/issues/370#issuecomment-3090793373

@absidue would this be accepted as a good solution, if it were to be more polished and tested?

**Comment by VewDev** [comment]
Note that I would consider self hosting a data storage server as "data is only stored locally" and that what I proposed (no default instance) would make it a opt-in feature, anyway.

**Comment by absidue** [comment]
@VewDev What I said above still applies, if we were to accept something onto FreeTube, it should either be directly between two FreeTube installs on the same local network or some local network only self-hosted sync server. One of the core reasons FreeTube exists is that it stores your data locally, so it would very obviously be out of scope to send user data to some third-party sync server and no not having a default instance is not good enough, because most users will just use the first cloud instance they can find. That is in the same bucket as the suggestions that we should add Google login support and the people that don't want to use it don't have to use it, a clear no.

**Comment by VewDev** [comment]
How would you limit self-hosted sync server to be local network only?

**Comment by absidue** [comment]
One option would be to ensure that it is single user only, just like FreeTube or have FreeTube refuse to use any sync server not on the local network.

**Comment by absidue** [comment]
Syncing directly between two FreeTube installs on the same network, without any extra servers in the middle, would be the option that fits FreeTube's principals the closest anyway.

**Comment by absidue** [comment]
@VewDev If you want to use something where your data is stored on a central server at all times, may I suggest that you use one of the many projects that does that such as Invidious, rather than suggesting that a project that was designed to not store your data on a central server, should start storing people's data on a central server.

**Comment by Nama** [comment]
I use [syncthing](https://syncthing.net/) for this.

**Comment by Baptiste81** [comment]
I implemented a feature that use a json file in user directory which point to a location. Then Freetube reads that json file and read/write databases (settings, subscriptions, history, ...) from that location. If not clear here is an example : 

Freetube ---> <user_dir>/freetube-config.jsn ----> read/write <custom_dir>/history.db

It allows to save databases into a directory on cloud provider (owncloud, nextcloud, dropbox, pcloud, ...) that is always sync between your machines.

Feature proposal is #9727 PR is #9728
